src/cpp/core/system/LinuxResources.cpp

```diff
@@ -0,0 +1,431 @@
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace core {
+namespace system {
+namespace {
+
+/**
+ * LinuxMemoryProvider is an abstract class that provides memory usage stats on
+ * Linux. As there are several ways of limiting and measuring memory usage on
+ * Linux, we figure out the most appropriate one at runtime and create the 
+ * corresponding provider (see getMemoryProvider factory)
+ */
+class LinuxMemoryProvider
+{
+public:
+
+   virtual Error getMemoryUsed(long *pUsedKb, MemoryProvider *pProvider) = 0;
+
+   virtual Error getTotalMemory(long *pTotalKb, MemoryProvider *pProvider) = 0;
+
+   Error getProcessMemoryUsed(long *pUsedKb, MemoryProvider *pProvider)
+   {
+      long size = 0;
+      long resident = 0;
+      try 
+      {
+         std::ifstream statm("/proc/self/statm");
+         statm >> size >> resident;
+         statm.close();
+      }
+      catch (...)
+      {
+         Error error = systemError(boost::system::errc::no_such_file_or_directory, 
+               "Could not read process memory stats from /proc/self/statm", 
+               ERROR_LOCATION);
+         return error;
+      }
+
+      long pageKib = ::sysconf(_SC_PAGE_SIZE) / 1024;
+      *pUsedKb = resident * pageKib;
+      *pProvider = MemoryProviderLinuxProcFs;
+      return Success();
+   }
+};
+
+/**
+ * MemInfoMemoryProvider is a class supplying Linux memory statistics using the
+ * contents of the file /proc/meminfo.
+ */
+class MemInfoMemoryProvider : public LinuxMemoryProvider
+{
+public:
+   MemInfoMemoryProvider(): memTotal_(0)
+   {
+      // Read total memory only once (it doesn't change)
+      Error error = readMemInfoKey("MemTotal", &memTotal_);
+      if (error)
+      {
+         LOG_ERROR(error);
+      }
+   }
+
+   Error getMemoryUsed(long *pUsedKb, MemoryProvider *pProvider)
+   {
+      long availableKb = 0;
+      Error error = readMemInfoKey("MemAvailable", &availableKb);
+      if (error)
+      {
+         return error;
+      }
+      *pUsedKb = memTotal_ - availableKb;
+      *pProvider = MemoryProviderLinuxProcMeminfo;
+      return Success();
+   }
+
+   Error getTotalMemory(long *pTotalKb, MemoryProvider *pProvider)
+   {
+      *pTotalKb = memTotal_;
+      *pProvider = MemoryProviderLinuxProcMeminfo;
+      return Success();
+   }
+
+private:
+
+   // Parses /proc/meminfo to look up specific memory stats.
+   Error readMemInfoKey(const std::string& key, long* pValue)
+   {
+      // /proc/meminfo contains lines that look like this:
+      //
+      // MemTotal: 8124360 kB
+      
+      // Open the file and prepare to read it
+      FilePath memInfoFile("/proc/meminfo");
+      std::shared_ptr<std::istream> pMemStream;
+      Error error = memInfoFile.openForRead(pMemStream);
+      if (error)
+      {
+         return error;
+      }
+
+      // Read one line at a time, looking for the key
+      while (!pMemStream->eof())
+      {
+         std::string memLine;
+         std::getline(*pMemStream, memLine);
```

Comment:
I don't think that can happen because it's a virtual file, but it's easier to add a guard than to try to prove we don't need one. I've added a try/catch around this line.